### PR TITLE
sched/signal: add spinlock to g_sigfreeaction

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary
sched/signal: add spinlock to g_sigfreeaction

To avoid nxsig_alloc_action() & nxsig_release_action() competition

## Impact

sig_action

### CODE REVIEW

File: sched/signal/sig_action.c [hunk 1]
```diff
@@ -53,10 +53,13 @@
 static FAR sigactq_t *nxsig_alloc_action(void)
 {
   FAR sigactq_t *sigact;
+  irqstate_t flags;
 
   /* Try to get the signal action structure from the free list */
 
+  flags = spin_lock_irqsave(NULL);
```
Comment (Ouss4) [inline sched/signal/sig_action.c:60]: Do we need the global IRQ lock here?  I think a private lock would work fine.

Comment (Ouss4) [reply sched/signal/sig_action.c:60]: Hmmm... I'm not sure, `g_sigfreeacation` is global and accessed elsewhere.  Can we add the same lock for line 67?

Comment (Ouss4) [reply sched/signal/sig_action.c:60]: BTW can't we move `nxsig_alloc_actionblock` to `sig_action.c`?

Comment (GUIDINGLI) [reply sched/signal/sig_action.c:103]: done

**Comment by Ouss4** [comment]
@xiaoxiang781216 @GUIDINGLI How about using a private lock here instead of the global one? cc @masayuki2009

**Comment by masayuki2009** [comment]
>@xiaoxiang781216 @GUIDINGLI How about using a private lock here instead of the global one? cc @masayuki2009

Though I think we should use a private lock as much as possible, let me check this PR with spresense.

File: sched/signal/sig_action.c [hunk 2]
```diff
@@ -38,10 +38,55 @@
...
+
+static void nxsig_alloc_actionblock(void)
+{
+  FAR sigactq_t *sigact;
+  irqstate_t flags;
+  int i;
+
+  /* Allocate a block of signal actions */
+
+  g_sigactionalloc =
```
Comment (xiaoxiang781216) [inline sched/signal/sig_action.c:72]: let's remove g_sigactionalloc, which isn't really needed at all.

**Comment by xiaoxiang781216** [comment]
@masayuki2009 do you have any progress?

**Comment by masayuki2009** [comment]
>@masayuki2009 do you have any progress?

I confirmed this PR works with spresense SMP configurations.
But I prefer to use a private spinlock instead of the global one to improve performance.

**Review by masayuki2009 — APPROVED** [review]
>I confirmed this PR works with spresense SMP configurations.
>But I prefer to use a private spinlock instead of the global one to improve performance.

Let me merge this PR so that I can create a new PR for it.

**Pull request merged**